infer/src/bufferoverrun/bufferOverrunAnalysis.ml

```diff
@@ -80,7 +80,34 @@ module TransferFunctions = struct
         Option.value_map (Dom.Mem.find_opt loc callee_exit_mem) ~default:acc ~f:(fun v ->
             let locs = PowLoc.subst_loc loc eval_locpath in
             let v = Dom.Val.subst v eval_sym_trace location in
-            PowLoc.fold (fun loc acc -> Dom.Mem.add_heap loc v acc) locs acc )
+            (* Always do strong updates if the following two conditions hold
+               1) Unique allocsites are asumed: a single allocsite doesn't represent multiple
+                  objects. For example:
+
+                 Arr1 = createArray (length1); – creates an array of length length1
+                 Arr2 = createArray (length2); – creates an array of length length2
+
+                 If it can happen that Arr1 and Arr2 is represented by the same allocsite,
+                 weak update has to be always performed.
+
+                 Currently, BO doesn't ensure that this hold, but frontend can ensure it
+                 by copying an allocsite imported from the callee to new one after the call.
```

Comment:
> BO could resolve this by adding an identification of the caller to the allocsite when importing it fom the callee to the caller.

I see. You are giving context sensitivity only for the allocsite locations. Very nice! Maybe I missed the code that replaces `caller_pname` when importing/instantiating callee's summary.

Reply:
> I see. You are giving context sensitivity only for the allocsite locations.

Exactly. Looking at the name of the option and the comment, I see where the confusion came from. I renamed the option to `bo_context_sensitive_allocsites` and rewrote the comment.

Reply:
> You are giving context sensitivity only for the allocsite locations. Very nice! Maybe I missed the code that replaces caller_pname when importing/instantiating callee's summary.

Just to clarify - I think that there is still some confusion. I didn't add the context sensitivity you probably think :(. `caller_pname` is never substituted so it is always the name of the caller of the builtin calling malloc. This is to solve the problem that otherwise, we can have two identical allocsites created in two unrelated functions (these functions would call the same builtin to create the allocsite so `proc_name` would be the same and it could also happen that `node_hash` would be the same).

For Ada frontend and return values of functions, this is just enough because it is explicitly copying return values of functions after the call so we can then assume context-sensitive allocsites and do strong updates. However, in general, this assumption doesn't hold for all the frontends and weak updates need to be performed. In the comment, I gave an example of such situation where a strong update would be incorrect and noted that this could be overcomed by adding more context sensitivity to allocsites.

Reply:
> Just to clarify - I think that there is still some confusion. I didn't add the context sensitivity you probably think  :(. caller_pname is never substituted

Thank you. It is now more clear where was the problem and how you addressed it, thanks to the discussions.